--- mime.types.orig	2012-04-06 18:07:19.000000000 +0200
+++ mime.types	2012-04-06 18:10:14.000000000 +0200
@@ -89,7 +89,6 @@
 application/iotp				
 application/ipp				
 application/isup				
-application/java-archive				jar
 application/java-serialized-object				ser
 application/java-vm				class
 application/javascript				js
@@ -782,6 +781,7 @@
 application/wsdl+xml				wsdl
 application/wspolicy+xml				wspolicy
 application/x-123				wk
+application/x-7z-compressed			7z
 application/x-abiword				abw
 application/x-ace-compressed				ace
 application/x-apple-diskimage				dmg
@@ -790,8 +790,9 @@
 application/x-authorware-seg				aas
 application/x-bcpio				bcpio
 application/x-bittorrent				torrent
-application/x-bzip				bz
-application/x-bzip2				boz bz2
+application/x-bzip				bz 
+application/x-bzip2				boz bz2 
+application/x-bzip-compressed-tar		tbz tbz2
 application/x-cdf				cdf
 application/x-cdlink				vcd
 application/x-chat				chat
@@ -799,7 +800,7 @@
 application/x-compress				
 application/x-cpio				cpio
 application/x-csh				csh
-application/x-debian-package				deb udeb
+application/x-deb				deb udeb
 application/x-director				cct cst cxt dcr dir dxr fgd swa w3d
 application/x-dms				dms
 application/x-doom				wad
@@ -828,14 +829,15 @@
 application/x-gnumeric				gnumeric
 application/x-go-sgf				sgf
 application/x-graphing-calculator				gcf
-application/x-gtar				gtar taz tgz
-application/x-gzip				
+application/x-gtar				gtar taz
+application/x-gzip				gz
 application/x-hdf				hdf
 application/x-ica				ica
 application/x-internet-signup				ins isp
 application/x-iphone				iii
 application/x-iso9660-image				iso
 application/x-java-jnlp-file				jnlp
+application/x-java-archive				jar
 application/x-javascript				js
 application/x-jmol				jmz
 application/x-kchart				chrt
@@ -847,6 +849,7 @@
 application/x-latex				latex
 application/x-lha				lha
 application/x-lzh				lzh
+application/x-lzma				xz lzma
 application/x-lzx				lzx
 application/x-maker				book fb fbdoc fm frame frm maker
 application/x-mif				mif
@@ -880,7 +883,7 @@
 application/x-pkcs7-crl				crl
 application/x-python-code				pyc pyo
 application/x-quicktimeplayer				qtl
-application/x-rar-compressed				rar
+application/x-rar				rar
 application/x-redhat-package-manager				rpm
 application/x-sh				sh
 application/x-shar				shar
