--- ./CMakeLists.txt.orig	2014-04-14 10:35:10.000000000 +0200
+++ ./CMakeLists.txt	2014-04-14 10:36:57.479951904 +0200
@@ -116,6 +116,9 @@ if(ENABLE_GTK2)
   find_package(PkgConfig REQUIRED)
   unset(__pkg_config_checked_GTK2 CACHE)
   pkg_check_modules(GTK2 REQUIRED gtk+-2.0)
+  STRING(REGEX REPLACE "/usr" "${CMAKE_INSTALL_PREFIX}" GTK2_THEME_PREFIX "${GTK2_THEME_PREFIX}")
+  STRING(REGEX REPLACE "/usr" "${CMAKE_INSTALL_PREFIX}" GTK2_LIBDIR "${GTK2_LIBDIR}")
+  STRING(REGEX REPLACE "/usr" "${CMAKE_INSTALL_PREFIX}" GTK2_PREFIX "${GTK2_PREFIX}")
   if(GTK2_THEME_PREFIX)
     set(GTK2_PREFIX "${GTK2_THEME_PREFIX}")
   elseif(NOT GTK2_PREFIX)
--- ./qt4/CMakeLists.txt.orig	2014-01-25 21:13:54.000000000 +0100
+++ ./qt4/CMakeLists.txt	2014-04-14 10:36:08.980953609 +0200
@@ -22,6 +22,7 @@ if(ENABLE_QT4)
       endif()
     endif()
   else()
+    STRING(REGEX REPLACE "/usr" "${CMAKE_INSTALL_PREFIX}" QT_PLUGINS_DIR "${QT_PLUGINS_DIR}")
     set(QTCURVE_STYLE_DIR "${QT_PLUGINS_DIR}/styles/")
   endif()
   add_definitions(${QT_DEFINITIONS} -DQTC_UTILS_QT4)
