
--- libkgreeter-4.11.19/CMakeLists.txt.orig	2016-02-23 15:12:24.527645125 +0100
+++ libkgreeter-4.11.19/CMakeLists.txt	2016-02-23 15:12:26.599645092 +0100
@@ -198,7 +198,7 @@
   configure_file(config-unix.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config-unix.h )
   configure_file(config-X11.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config-X11.h )
   if(NOT CMAKE_INSTALL_PREFIX STREQUAL "/usr")
-    set(EXPORT_XCURSOR_PATH "XCURSOR_PATH=${CMAKE_INSTALL_PREFIX}/share/icons:$XCURSOR_PATH\":~/.icons:/usr/share/icons:/usr/share/pixmaps:/usr/X11R6/lib/X11/icons\"; export XCURSOR_PATH")
+    set(EXPORT_XCURSOR_PATH "XCURSOR_PATH=${CMAKE_INSTALL_PREFIX}/share/icons:$XCURSOR_PATH\":~/.icons:/usr/local/share/icons:/usr/local/share/pixmaps:/usr/X11R6/lib/X11/icons\"; export XCURSOR_PATH")
   endif(NOT CMAKE_INSTALL_PREFIX STREQUAL "/usr")
 #DONOTINSTALL   configure_file(startkde.cmake ${CMAKE_CURRENT_BINARY_DIR}/startkde  @ONLY)
 endif(NOT WIN32)

