Index: perl-5.16.3/vms/vms.c
===================================================================
--- perl-5.16.3.orig/vms/vms.c
+++ perl-5.16.3/vms/vms.c
@@ -1370,7 +1370,9 @@ prime_env_iter(void)
     if (!str$case_blind_compare(env_tables[i],&crtlenv)) {
       char *start;
       int j;
-      for (j = 0; environ[j]; j++) { 
+      for (j = 0; environ[j]; j++); 
+      /*Start at the end so if there is duplicate we keep the first one*/
+      for (j--; j>=0; j--) { 
         if (!(start = strchr(environ[j],'='))) {
           if (ckWARN(WARN_INTERNAL)) 
             Perl_warner(aTHX_ packWARN(WARN_INTERNAL),"Ill-formed CRTL environ value \"%s\"\n",environ[j]);
