Index: shadow-4.1.5.1/src/Makefile.am
===================================================================
--- shadow-4.1.5.1.orig/src/Makefile.am
+++ shadow-4.1.5.1/src/Makefile.am
@@ -49,19 +49,18 @@ usbin_PROGRAMS = \
 noinst_PROGRAMS = id sulogin
 
 suidbins       = su
-suidubins      = chage chfn chsh expiry gpasswd newgrp passwd
+suidubins      = chage chfn chsh expiry gpasswd newgrp
 if ACCT_TOOLS_SETUID
 	suidubins += chage chgpasswd chpasswd groupadd groupdel groupmod newusers useradd userdel usermod
 endif
 
 if WITH_TCB
-suidubins -= passwd
 shadowsgidubins = passwd
+else
+suidubins += passwd
 endif
 
-LDADD          = $(INTLLIBS) \
-		 $(LIBTCB) \
-		 $(top_builddir)/libmisc/libmisc.a \
+LDADD          = $(top_builddir)/libmisc/libmisc.a \
 		 $(top_builddir)/lib/libshadow.la
 AM_CPPFLAGS    = -DLOCALEDIR=\"$(datadir)/locale\"
 
@@ -77,39 +76,39 @@ else
 LIBCRYPT_NOPAM = $(LIBCRYPT)
 endif
 
-chage_LDADD    = $(LDADD) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
-chfn_LDADD     = $(LDADD) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
-chgpasswd_LDADD = $(LDADD) $(LIBPAM_SUID) $(LIBSELINUX) $(LIBCRYPT)
-chsh_LDADD     = $(LDADD) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
-chpasswd_LDADD = $(LDADD) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT)
-gpasswd_LDADD  = $(LDADD) $(LIBAUDIT) $(LIBSELINUX) $(LIBCRYPT)
-groupadd_LDADD = $(LDADD) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
-groupdel_LDADD = $(LDADD) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
-groupmems_LDADD = $(LDADD) $(LIBPAM) $(LIBSELINUX)
-groupmod_LDADD = $(LDADD) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
-grpck_LDADD    = $(LDADD) $(LIBSELINUX)
-grpconv_LDADD  = $(LDADD) $(LIBSELINUX)
-grpunconv_LDADD = $(LDADD) $(LIBSELINUX)
+chage_LDADD    = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
+chfn_LDADD     = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
+chgpasswd_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBSELINUX) $(LIBCRYPT) -lpam_userpass
+chsh_LDADD     = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
+chpasswd_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT) -lpam_userpass
+gpasswd_LDADD  = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBAUDIT) $(LIBSELINUX) $(LIBCRYPT)
+groupadd_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
+groupdel_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
+groupmems_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBSELINUX)
+groupmod_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX)
+grpck_LDADD    = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBSELINUX)
+grpconv_LDADD  = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBSELINUX)
+grpunconv_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBSELINUX)
 login_SOURCES  = \
 	login.c \
 	login_nopam.c
-login_LDADD    = $(LDADD) $(LIBPAM) $(LIBAUDIT) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
-newgrp_LDADD   = $(LDADD) $(LIBAUDIT) $(LIBCRYPT)
-newusers_LDADD = $(LDADD) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT)
-nologin_LDADD  =
-passwd_LDADD   = $(LDADD) $(LIBPAM) $(LIBCRACK) $(LIBAUDIT) $(LIBSELINUX) $(LIBCRYPT_NOPAM)
-pwck_LDADD     = $(LDADD) $(LIBSELINUX)
-pwconv_LDADD   = $(LDADD) $(LIBSELINUX)
-pwunconv_LDADD = $(LDADD) $(LIBSELINUX)
+login_LDADD    = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBAUDIT) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
+newgrp_LDADD   = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBAUDIT) $(LIBCRYPT)
+newusers_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT) -lpam_userpass
+nologin_LDADD  =
+passwd_LDADD   = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBCRACK) $(LIBAUDIT) $(LIBSELINUX) $(LIBCRYPT_NOPAM)
+pwck_LDADD     = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBSELINUX)
+pwconv_LDADD   = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBSELINUX)
+pwunconv_LDADD = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBSELINUX)
 su_SOURCES     = \
 	su.c \
 	suauth.c
-su_LDADD       = $(LDADD) $(LIBPAM) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
-sulogin_LDADD  = $(LDADD) $(LIBCRYPT)
-useradd_LDADD  = $(LDADD) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX) $(LIBSEMANAGE) $(LIBACL) $(LIBATTR)
-userdel_LDADD  = $(LDADD) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX) $(LIBSEMANAGE)
-usermod_LDADD  = $(LDADD) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX) $(LIBSEMANAGE) $(LIBACL) $(LIBATTR)
-vipw_LDADD     = $(LDADD) $(LIBSELINUX)
+su_LDADD       = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
+sulogin_LDADD  = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBCRYPT)
+useradd_LDADD  = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX) $(LIBSEMANAGE) $(LIBACL) $(LIBATTR)
+userdel_LDADD  = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX) $(LIBSEMANAGE)
+usermod_LDADD  = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBPAM_SUID) $(LIBAUDIT) $(LIBSELINUX) $(LIBSEMANAGE) $(LIBACL) $(LIBATTR)
+vipw_LDADD     = $(LDADD) $(INTLLIBS) $(LIBTCB) $(LIBSELINUX)
 
 install-am: all-am
 	$(MAKE) $(AM_MAKEFLAGS) install-exec-am install-data-am
