--- src/files.c.bkp	2009-03-19 14:40:31.179629618 +0100
+++ src/files.c	2009-03-19 14:42:31.413287220 +0100
@@ -1572,7 +1572,8 @@ bool write_file(const char *name, FILE *
 	}
 
 	if (f_open == NULL) {
-	    fd_source = open(realname, O_RDONLY | O_CREAT);
+	    fd_source = open(realname, O_RDONLY | O_CREAT, 
+			S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH);
 
 	    if (fd_source != -1) {
 		f_source = fdopen(fd_source, "rb");
@@ -1691,7 +1692,8 @@ bool write_file(const char *name, FILE *
 	int fd_source;
 	FILE *f_source = NULL;
 
-	fd_source = open(tempname, O_RDONLY | O_CREAT);
+	fd_source = open(tempname, O_RDONLY | O_CREAT,
+			S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH);
 
 	if (fd_source != -1) {
 	    f_source = fdopen(fd_source, "rb");
