--- configure.ac.orig	2012-02-13 15:26:31.000000000 +0100
+++ configure.ac	2012-02-13 15:26:45.000000000 +0100
@@ -81,11 +81,6 @@ sdkdir=$(pkg-config --variable=sdkdir xo
 # Checks for libraries.
 PKG_CHECK_MODULES([PCIACCESS], [pciaccess >= 0.10])
 
-PKG_CHECK_MODULES(LIBUDEV, [libudev], [LIBUDEV=yes], [LIBUDEV=no])
-if test "x$LIBUDEV" = xyes; then
-	AC_DEFINE(HAVE_LIBUDEV, 1, [libudev support])
-fi
-AM_CONDITIONAL(LIBUDEV, [ test "x$LIBUDEV" = "xyes" ] )
 AC_SUBST([LIBUDEV_CFLAGS])
 AC_SUBST([LIBUDEV_LIBS])
 
