Index: thunderbird-52.6.0/mail/base/content/specialTabs.js
===================================================================
--- thunderbird-52.6.0.orig/mail/base/content/specialTabs.js
+++ thunderbird-52.6.0/mail/base/content/specialTabs.js
@@ -879,6 +879,10 @@ var specialTabs = {
     Services.prefs.clearUserPref(kTelemetryEnabled);
     Services.prefs.clearUserPref(kTelemetryRejected);
 
+    // Pour CLIP : on fait comme si l'utilisateur avait clique sur non
+    Services.prefs.setBoolPref(kTelemetryRejected, true);
+
+    /*
     var buttons = [
       {
         label:     telemetryBundle.get("telemetryYesButtonLabel"),
@@ -918,6 +922,7 @@ var specialTabs = {
 
     let description = notification.ownerDocument.getAnonymousElementByAttribute(notification, "anonid", "messageText");
     description.appendChild(link);
+    */
   },
   /**
    * Looks at the existing prefs and determines if we should show about:rights
