diff -Naur Linux-PAM-1.1.0-orig/modules/pam_wheel/pam_wheel.c Linux-PAM-1.1.0/modules/pam_wheel/pam_wheel.c
--- Linux-PAM-1.1.0-orig/modules/pam_wheel/pam_wheel.c	2007-11-06 16:24:26.000000000 +0100
+++ Linux-PAM-1.1.0/modules/pam_wheel/pam_wheel.c	2010-11-08 08:51:31.000000000 +0100
@@ -64,6 +64,7 @@
 #define PAM_TRUST_ARG       0x0004
 #define PAM_DENY_ARG        0x0010
 #define PAM_ROOT_ONLY_ARG   0x0020
+#define PAM_USE_LOGIN_ARG   0x0040
 
 static int
 _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
@@ -82,6 +83,8 @@
                ctrl |= PAM_DEBUG_ARG;
           else if (!strcmp(*argv,"use_uid"))
                ctrl |= PAM_USE_UID_ARG;
+          else if (!strcmp(*argv,"use_login"))
+               ctrl |= PAM_USE_LOGIN_ARG;
           else if (!strcmp(*argv,"trust"))
                ctrl |= PAM_TRUST_ARG;
           else if (!strcmp(*argv,"deny"))
@@ -138,6 +141,23 @@
 	    return PAM_SERVICE_ERR;
 	}
 	fromsu = tpwd->pw_name;
+    } else if (ctrl & PAM_USE_LOGIN_ARG) {
+      fromsu = pwd->pw_name;
+      if (!fromsu) {
+	if (ctrl & PAM_DEBUG_ARG) {
+	  pam_syslog(pamh, LOG_NOTICE, "no username");
+	}
+	return PAM_SERVICE_ERR;
+      }
+      
+      tpwd = pam_modutil_getpwnam (pamh, fromsu);
+      if (!tpwd) {
+	if (ctrl & PAM_DEBUG_ARG) {
+	  pam_syslog(pamh, LOG_NOTICE, "getpwnam : error");
+	}
+	return PAM_SERVICE_ERR;
+      }
+
     } else {
 	fromsu = pam_modutil_getlogin(pamh);
 	if (fromsu) {
@@ -145,7 +165,7 @@
 	}
 	if (!fromsu || !tpwd) {
 	    if (ctrl & PAM_DEBUG_ARG) {
-		pam_syslog(pamh, LOG_NOTICE, "who is running me ?!");
+		pam_syslog(pamh, LOG_NOTICE, "who is running me ?");
 	    }
 	    return PAM_SERVICE_ERR;
 	}
@@ -189,7 +209,9 @@
      */
 
     if (is_on_list(grp->gr_mem, fromsu) || (tpwd->pw_gid == grp->gr_gid)) {
-
+	if (ctrl & PAM_DEBUG_ARG) {
+	  pam_syslog(pamh, LOG_NOTICE, "user %s is in group %d\n", fromsu, grp->gr_gid);
+	}
 	if (ctrl & PAM_DENY_ARG) {
 	    retval = PAM_PERM_DENIED;
 
@@ -201,6 +223,9 @@
 	}
 
     } else {
+	if (ctrl & PAM_DEBUG_ARG) {
+	  pam_syslog(pamh, LOG_NOTICE, "user %s is NOT in group %d\n", fromsu, grp->gr_gid);
+	}
 
 	if (ctrl & PAM_DENY_ARG) {
 
