Index: firefox-52.3.0esr/old-configure.in
===================================================================
--- firefox-52.3.0esr.orig/old-configure.in
+++ firefox-52.3.0esr/old-configure.in
@@ -5521,23 +5521,23 @@ AC_SUBST(MOZ_SOURCE_REPO)
 AC_SUBST(MOZ_SOURCE_CHANGESET)
 AC_SUBST(MOZ_INCLUDE_SOURCE_INFO)
 
-if test "$MOZ_TELEMETRY_REPORTING"; then
-    AC_DEFINE(MOZ_TELEMETRY_REPORTING)
+dnl if test "$MOZ_TELEMETRY_REPORTING"; then
+dnl    AC_DEFINE(MOZ_TELEMETRY_REPORTING)
 
-    # Enable Telemetry by default for nightly and aurora channels
-    if test -z "$RELEASE_OR_BETA"; then
-      AC_DEFINE(MOZ_TELEMETRY_ON_BY_DEFAULT)
-    fi
-fi
+dnl    # Enable Telemetry by default for nightly and aurora channels
+dnl    if test -z "$RELEASE_OR_BETA"; then
+dnl      AC_DEFINE(MOZ_TELEMETRY_ON_BY_DEFAULT)
+dnl    fi
+dnl fi
 
 dnl If we have any service that uploads data (and requires data submission
 dnl policy alert), set MOZ_DATA_REPORTING.
 dnl We need SUBST for build system and DEFINE for xul preprocessor.
-if test -n "$MOZ_TELEMETRY_REPORTING" || test -n "$MOZ_SERVICES_HEALTHREPORT" || test -n "$MOZ_CRASHREPORTER"; then
-  MOZ_DATA_REPORTING=1
-  AC_DEFINE(MOZ_DATA_REPORTING)
-  AC_SUBST(MOZ_DATA_REPORTING)
-fi
+dnl if test -n "$MOZ_TELEMETRY_REPORTING" || test -n "$MOZ_SERVICES_HEALTHREPORT" || test -n "$MOZ_CRASHREPORTER"; then
+dnl   MOZ_DATA_REPORTING=1
+dnl   AC_DEFINE(MOZ_DATA_REPORTING)
+dnl   AC_SUBST(MOZ_DATA_REPORTING)
+dnl fi
 
 dnl win32 options
 AC_SUBST(WIN32_REDIST_DIR)
